### PR TITLE
Introduces working Status Displays

### LINKED ISSUE #1742: Windows block vision for objects on top of walls (APCs only?)
# Description
image
It's very noticeable on the APC in front of the captain's office, when walking on the room to the right it disappears from your vision, I cant seem to find this error in other part of the station

## Category
BUG

## Steps to Reproduce
Please enter the steps to reproduce the bug or behaviour:
1.Walk in front of the captains office
2.Watch the APC as you move back and forwards on the room to the right
3.Notice how it disappears even through there's a window

### LINKED ISSUE #1773: There's an APC on top of a window
# Description
image
This APC located in front of the captain's office is on top of a window, APCs should only be on top of walls

## Category
BUG

## Steps to Reproduce
Please enter the steps to reproduce the bug or behaviour:
1.Start the game and head towards the room in front of the captains office
2.Notice the APC on top of a window

### LINKED ISSUE #2170: Tile-related stuff like blood, fire and Wet floors have missing sprites
# Description
Even though they are set on the prefabs they still don't Show up
## Category
BUG 

## Steps to Reproduce
Please enter the steps to reproduce the bug or behaviour:
1. Use mop and floor observe lack of wet
2. Walkover "wet" spot and trip to prove it is there
3. B 
4. Shoot a station pet, Right click observe the Missing sprites for the blood

### PR DESCRIPTION
**Pull request by fomalsd**
### Purpose
Fixes #2170 
Fixes #1773 
Fixes #1742 
Partial fix for #2148
Adds Status Displays (not mapped yet) that show short text messages broadcasted to currently selected channel – up to 2 pages, 18 characters each.
Comms console allows composing and broadcasting custom messages to displays with active `Command` (default) channel.
`Escape` channel is a priority one and will overtake other channels – it shows primary escape shuttle ETA to station, time until shuttle departure and ETA from station to centcom
image

bugs: apparently fov masked material can't be applied to text, therefore text is visible through fov 

### Open Questions and Pre-Merge TODOs

- [x]  This fix is tested on the same branch it is PR'ed to.
- [x]  I correctly commented my code
- [x]  My code is indented with tabs and not spaces
- [x]  This PR does not include any unnecessary .meta, .prefab or .unity (scene) changes
- [x]  This PR does not bring up any new compile errors
- [x]  This PR has been tested in editor
- [x]  This PR has been tested in multiplayer (with 2 clients and late joiners, if applicable)

### CODE REVIEW

**Comment by fomalsd** [comment]
to broadcast text use 
`GameManager.Instance.CentComm.OnStatusDisplayUpdate.Invoke( channel, text );`

**Comment by fomalsd** [comment]
there's some Mirror stuff going on, likely caused by me renaming CellTimers into StatusDisplays, so I'll have to remap cell ones, add some more and include outpost scene
```
OnDeserialize failed for: object=CellTimer (1) component=StatusDisplay sceneId=BD781D801B6A592D length=2. Possible Reasons:
  * Do StatusDisplay's OnSerialize and OnDeserialize calls write the same amount of data(2 bytes)? 
  * Was there an exception in StatusDisplay's OnSerialize/OnDeserialize code?
  * Are the server and client the exact same project?
  * Maybe this OnDeserialize call was meant for another GameObject? The sceneIds can easily get out of sync if the Hierarchy was modified only in the client OR the server. Try rebuilding both.

System.NullReferenceException: Object reference not set to an instance of an object
  at StatusDisplay.SetStatusText (System.String newText) [0x00001] in C:\homeproj\unitystation\UnityProject\Assets\Scripts\Objects\StatusDisplay.cs:59 
  at StatusDisplay.OnDeserialize (Mirror.NetworkReader reader, System.Boolean initialState) [0x00014] in <2fdc178064c74d08a8f97b31db759f9e>:0 
  at Mirror.NetworkIdentity.OnDeserializeSafely (Mirror.NetworkBehaviour comp, Mirror.NetworkReader reader, System.Boolean initialState) [0x00081] in C:\homeproj\unitystation\UnityProject\Assets\Mirror\Runtime\NetworkIdentity.cs:872 
UnityEngine.Debug:LogError(Object)
Mirror.NetworkIdentity:OnDeserializeSafely(NetworkBehaviour, NetworkReader, Boolean) (at Assets/Mirror/Runtime/NetworkIdentity.cs:877)
Mirror.NetworkIdentity:OnDeserializeAllSafely(NetworkReader, Boolean) (at Assets/Mirror/Runtime/NetworkIdentity.cs:915)
Mirror.NetworkIdentity:OnUpdateVars(NetworkReader, Boolean) (at Assets/Mirror/Runtime/NetworkIdentity.cs:986)
Mirror.ClientScene:ApplySpawnPayload(NetworkIdentity, Vector3, Quaternion, Vector3, ArraySegment`1, UInt32) (at Assets/Mirror/Runtime/ClientScene.cs:450)
Mirror.ClientScene:OnSpawnSceneObject(NetworkConnection, SpawnSceneObjectMessage) (at Assets/Mirror/Runtime/ClientScene.cs:567)
Mirror.<>c__DisplayClass5_0`1:<MessageHandler>b__0(NetworkMessage) (at Assets/Mirror/Runtime/MessagePacker.cs:133)
Mirror.NetworkConnection:InvokeHandler(Int32, NetworkReader) (at Assets/Mirror/Runtime/NetworkConnection.cs:293)
Mirror.NetworkConnection:TransportReceive(ArraySegment`1) (at Assets/Mirror/Runtime/NetworkConnection.cs:334)
Mirror.NetworkClient:OnDataReceived(ArraySegment`1) (at Assets/Mirror/Runtime/NetworkClient.cs:165)
UnityEngine.Events.UnityEvent`1:Invoke(ArraySegment`1)
Mirror.TelepathyTransport:ProcessClientMessage() (at Assets/Mirror/Runtime/Transport/TelepathyTransport.cs:65)
Mirror.TelepathyTransport:LateUpdate() (at Assets/Mirror/Runtime/Transport/TelepathyTransport.cs:92)
```

**Comment by fomalsd** [comment]
Added a bunch of fixes. Someone please check this out in multiplayer: I still get this mirror nagging, but everything seems to work fine:
```
OnDeserialize failed for: object=StatusDisplay (1) component=StatusDisplay sceneId=BD781D8099807869 length=2. Possible Reasons:
  * Do StatusDisplay's OnSerialize and OnDeserialize calls write the same amount of data(2 bytes)? 
  * Was there an exception in StatusDisplay's OnSerialize/OnDeserialize code?
  * Are the server and client the exact same project?
  * Maybe this OnDeserialize call was meant for another GameObject? The sceneIds can easily get out of sync if the Hierarchy was modified only in the client OR the server. Try rebuilding both.

System.NullReferenceException: Object reference not set to an instance of an object
  at StatusDisplay.SetStatusText (System.String newText) [0x00001] in C:\homeproj\unitystation\UnityProject\Assets\Scripts\Objects\StatusDisplay.cs:60 
  at StatusDisplay.OnDeserialize (Mirror.NetworkReader reader, System.Boolean initialState) [0x00014] in <25296ef7a1f44adf848564c2dc41e61e>:0 
  at Mirror.NetworkIdentity.OnDeserializeSafely (Mirror.NetworkBehaviour comp, Mirror.NetworkReader reader, System.Boolean initialState) [0x00081] in C:\homeproj\unitystation\UnityProject\Assets\Mirror\Runtime\NetworkIdentity.cs:872 
UnityEngine.Debug:LogError(Object)
Mirror.NetworkIdentity:OnDeserializeSafely(NetworkBehaviour, NetworkReader, Boolean) (at Assets/Mirror/Runtime/NetworkIdentity.cs:877)
Mirror.NetworkIdentity:OnDeserializeAllSafely(NetworkReader, Boolean) (at Assets/Mirror/Runtime/NetworkIdentity.cs:915)
Mirror.NetworkIdentity:OnUpdateVars(NetworkReader, Boolean) (at Assets/Mirror/Runtime/NetworkIdentity.cs:986)
Mirror.ClientScene:ApplySpawnPayload(NetworkIdentity, Vector3, Quaternion, Vector3, ArraySegment`1, UInt32) (at Assets/Mirror/Runtime/ClientScene.cs:450)
Mirror.ClientScene:OnSpawnSceneObject(NetworkConnection, SpawnSceneObjectMessage) (at Assets/Mirror/Runtime/ClientScene.cs:567)
Mirror.<>c__DisplayClass5_0`1:<MessageHandler>b__0(NetworkMessage) (at Assets/Mirror/Runtime/MessagePacker.cs:133)
Mirror.NetworkConnection:InvokeHandler(Int32, NetworkReader) (at Assets/Mirror/Runtime/NetworkConnection.cs:293)
Mirror.NetworkConnection:TransportReceive(ArraySegment`1) (at Assets/Mirror/Runtime/NetworkConnection.cs:334)
Mirror.NetworkClient:OnDataReceived(ArraySegment`1) (at Assets/Mirror/Runtime/NetworkClient.cs:165)
UnityEngine.Events.UnityEvent`1:Invoke(ArraySegment`1)
Mirror.TelepathyTransport:ProcessClientMessage() (at Assets/Mirror/Runtime/Transport/TelepathyTransport.cs:65)
Mirror.TelepathyTransport:LateUpdate() (at Assets/Mirror/Runtime/Transport/TelepathyTransport.cs:92)
```

**Comment by fomalsd** [comment]
I think it should be good now, following syncvar best practices got rid of mirror errors
Still, someone please test it in multiplayer as Unity 2019.2 can be deceiving

File: UnityProject/Assets/Scripts/Objects/StatusDisplay.cs [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+	/// <summary>
+	/// used for restoring selected channel message when priority message is gone
+	/// </summary>
+	private string cachedText;
+
+	#region syncvar boilerplate and init
+
+	public override void OnStartClient()
+	{
+		SyncStatusText();
```
Comment (DooblyNoobly) [inline UnityProject/Assets/Scripts/Objects/StatusDisplay.cs]: change this to SyncStatusText(statusText)

File: UnityProject/Assets/Scripts/Objects/StatusDisplay.cs [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+	}
+	
+	public override void OnStartServer()
+	{
+		SyncStatusText();
```
Comment (DooblyNoobly) [inline UnityProject/Assets/Scripts/Objects/StatusDisplay.cs]: Here too

File: UnityProject/Assets/Scripts/Objects/StatusDisplay.cs [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+	}
+
+	public void GoingOnStageServer( OnStageInfo info )
+	{
+		SyncStatusText();
```
Comment (DooblyNoobly) [inline UnityProject/Assets/Scripts/Objects/StatusDisplay.cs]: and here

File: UnityProject/Assets/Scripts/Objects/StatusDisplay.cs [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+		GameManager.Instance.CentComm.OnStatusDisplayUpdate.AddListener( OnTextBroadcastReceived() );
+	}
+	
+	#endregion
+
+	/// <summary>
+	/// SyncVar hook to show text on client.
+	/// Text should be 2 pages max
+	/// </summary>
+	private void SyncStatusText(string newText = null)
```
Comment (DooblyNoobly) [inline UnityProject/Assets/Scripts/Objects/StatusDisplay.cs]: get rid of default param. Just make it '(string newText)'

File: UnityProject/Assets/Scripts/Objects/StatusDisplay.cs [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+	{
+		if ( newText == null)
```
Comment (DooblyNoobly) [inline UnityProject/Assets/Scripts/Objects/StatusDisplay.cs]: remove this whole if statement and just assign the param to status text.

```cs
statusText = newText;
```

**Review by DooblyNoobly — CHANGES_REQUESTED** [review]
The status is not synced with late connecting players as the syncvar hook is set up incorrectly. See my comments on how to fix it.

File: UnityProject/Assets/Scripts/Objects/StatusDisplay.cs (hunk 1 above)
Comment (fomalsd) [reply UnityProject/Assets/Scripts/Objects/StatusDisplay.cs]: I tried it before and got NREs on clientside, I think statusText is null at this point

Comment (DooblyNoobly) [reply UnityProject/Assets/Scripts/Objects/StatusDisplay.cs]: Weird, I just changed the code to make sure it works and it was fine. This was client side too

**Comment by fomalsd** [comment]
Fixed

**Comment by DooblyNoobly** [comment]
Yep that works. Looks good

**Pull request merged**